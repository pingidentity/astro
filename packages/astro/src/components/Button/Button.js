import React, { forwardRef, useRef, useImperativeHandle } from 'react';
import PropTypes from 'prop-types';
import { mergeProps, useButton, useFocusRing } from 'react-aria';
import { Button as ThemeUIButton } from 'theme-ui';
import { useHover } from '@react-aria/interactions';

import { useStatusClasses, usePropWarning, useAriaLabelWarning, useDeprecationWarning } from '../../hooks';
import Loader from '../Loader';

/**
 * Buttons are used to trigger actions or events.
 * This component is based on the [Button - Theme-UI](https://theme-ui.com/components/button/)
 * and includes a variety of styles, such as primary action, secondary action, or warning.
 */

const Button = forwardRef((props, ref) => {
  const {
    className,
    isDisabled,
    isLoading,
    onHoverStart,
    onHoverChange,
    onHoverEnd,
    onPress,
    onPressStart,
    onPressEnd,
    onPressChange,
    onPressUp,
    children,
    variant,
    tabIndex,
    ...others
  } = props;
  const buttonRef = useRef();
  usePropWarning(props, 'disabled', 'isDisabled');
  /* istanbul ignore next */
  useImperativeHandle(ref, () => buttonRef.current);

  const { isFocusVisible, focusProps } = useFocusRing();
  const { buttonProps, isPressed } = useButton({
    elementType: 'button',
    ...props,
  }, buttonRef);
  const { hoverProps, isHovered } = useHover(props);
  const { classNames } = useStatusClasses(className, {
    isHovered,
    isPressed,
    isFocused: isFocusVisible,
    isDisabled,
  });

  useDeprecationWarning(
    'The "icon" variant for `Button` is deprecated in Astro-UI 1.0.0, use the `IconButton` component instead.',
    { isActive: props.variant === 'icon' },
  );

  useDeprecationWarning(
    'The "danger" variant for `Button` will be deprecated in Astro-UI 2.0.0, use the `critical` variant instead.',
    { isActive: props.variant === 'danger' },
  );

  useDeprecationWarning(
    'The "text" variant for `Button` will be deprecated in Astro-UI 2.0.0, use the `link` variant instead.',
    { isActive: props.variant === 'text' },
  );

  const ariaLabel = props['aria-label'];
  useAriaLabelWarning('Button', ariaLabel, variant === 'filter');

  return (
    <ThemeUIButton
      aria-label={ariaLabel}
      ref={buttonRef}
      className={classNames}
      role="button"
      tx="buttons"
      sx={isLoading && { display: 'flex', justifyContent: 'center', alignItems: 'center' }}
      variant={variant}
      {...others}
      {...mergeProps({ ...buttonProps, tabIndex }, hoverProps, focusProps)}
    >
      {isLoading ? <span style={{ visibility: 'hidden' }}>{children}</span> : children}
      {isLoading && <Loader size="0.5em" sx={{ position: 'absolute' }} />}
    </ThemeUIButton>
  );
});

Button.propTypes = {
  /** Defines a string value that labels the current element. */
  'aria-label': PropTypes.string,
  /** Whether the button is disabled. */
  isDisabled: PropTypes.bool,
  /** Shows loader instead of children */
  isLoading: PropTypes.bool,
  /**
   * Handler that is called when a hover interaction starts.
   * (e: HoverEvent) => void
   */
  onHoverStart: PropTypes.func,
  /**
   * Handler that is called when a hover interaction ends.
   * (e: HoverEvent) => void
   */
  onHoverEnd: PropTypes.func,
  /**
   * Handler that is called when the hover state changes.
   * (isHovering: boolean) => void
   */
  onHoverChange: PropTypes.func,
  /**
   * Handler that is called when the press is released over the target.
   * (e: PressEvent) => void
   */
  onPress: PropTypes.func,
  /**
   * Handler that is called when a press interaction starts.
   * (e: PressEvent) => void
   */
  onPressStart: PropTypes.func,
  /**
   * Handler that is called when a press interaction ends, either over the target or when the
   * pointer leaves the target.
   * (e: PressEvent) => void
   */
  onPressEnd: PropTypes.func,
  /**
   * Handler that is called when the press state changes.
   * (isPressed: boolean) => void
   */
  onPressChange: PropTypes.func,
  /**
   * Handler that is called when a press is released over the target, regardless of whether it
   * started on the target or not.
   * (e: PressEvent) => void
   */
  onPressUp: PropTypes.func,
  /** The styling variation of the button. */
  variant: PropTypes.string,
  /** The focus variation of the button. */
  tabIndex: PropTypes.number,
};

Button.defaultProps = {
  isDisabled: false,
  variant: 'default',
};

Button.displayName = 'Button';
export default Button;
