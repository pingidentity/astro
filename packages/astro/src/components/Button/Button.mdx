<Meta title="Components/Button/Button" />

# Button

Button is used primarily to trigger an action or event. It is built on top of
[Button from Theme-UI](https://theme-ui.com/components/button). The stylings of a button is different based on its purpose.
(Primary Action, Secondary Action, Danger Button)

### Required components 

- Button component can be used independently and does not require additional components.

### Best practices

Button should:
- be clearly and accurately labeled.

Button shouldnâ€™t:
- be used as links.
- override colors (accessibility purposes).
- use uppercase titles.

### Accessibility

This component should adhere to the [WAI-ARIA Button](https://www.w3.org/TR/wai-aria-practices-1.1/examples/button/button.html) accessibility guidelines.

#### Keyboard Navigation

- Buttons are focusable using the Tab key and can be selected using the Space or Enter key. 

#### Screen readers.

Button utilizes `aria-label` to give an accessible name to the component.

| Aria Attribute | Role | Component |
| ---- | ---- | ----------- |
| `aria-label` | button | Button |