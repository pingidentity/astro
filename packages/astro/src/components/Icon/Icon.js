import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import Box from '../Box';

/**
 * Basic icon component.
 * Icons will fill the given container, which is '1em' x '1em' by default, and will maintain their
 * aspect ratio.
 * Use SVGR to import SVG files as React components and pass them into this component, and you can
 * use the theme to color the icon. But ensure you pass `dimensions: false` in webpack config
 * (see [SVGR docs](https://react-svgr.com/docs/options/#dimensions)).
 * Otherwise icon will not be resizable by `size` prop.
 * Icons can also be used from the [Material UI Icon Library](https://materialdesignicons.com/).
 * Built on top of the [Box from Rebass](https://rebassjs.org/box) and uses the
 * available [props from Rebass](https://rebassjs.org/props/).
*/

const Icon = forwardRef((props, ref) => {
  const {
    color,
    icon: IconComponent,
    size,
    sx, // eslint-disable-line
  } = props;

  return (
    <Box
      as={IconComponent}
      ref={ref}
      {...props}
      sx={{ fill: color, size: `${size}px`, minWidth: size, ...sx }}
    />
  );
});

Icon.propTypes = {
  /** The icon to render. */
  icon: PropTypes.elementType.isRequired,
  /**
   * The size of the icon container. If given a number value, it will be converted to pixels.
   * Standard icon sizes are 15, 22, and 40.
  */
  size: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /** A theme-aware prop to set the icon's color. */
  color: PropTypes.string,
};

Icon.defaultProps = {
  size: '22',
};
Icon.displayName = 'Icon';

export default Icon;
