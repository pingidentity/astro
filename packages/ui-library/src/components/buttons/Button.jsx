import React from "react";
import PropTypes from "prop-types";
import classnames from "classnames";
import HelpHint from "../tooltips/HelpHint";
import EllipsisLoader from "../general/EllipsisLoader";
import { cannonballChangeWarning } from "../../util/DeprecationUtils";
import { flagsPropType } from "../../util/FlagUtils";

/**
 * @class Button
 * @desc button component
 *
 * @param {("primary" | "secondary" | "standard" | "success" | "cancel" | "danger")} [type]
 *      CSS class applied to html.
 * @param {string} [alignInputs]
 *     When true, aligns a  button with text fields in the same row
 * @param {string} [className]
 *     Extra CSS class(s) applied to the top-level HTML container.
 * @param {string} [data-id="button"]
 *     Defines the "data-id" for top-level HTML container.
 * @param {string} [iconName]
 *     The name and CSS of the icon.
 * @param {function} [onClick]
 *     Click handler.
 * @param {string} [label]
 *     Html name of the button.
 * @param {boolean} [disabled=false]
 *     Button will not function when true.
 * @param {string} [disabledText]
 *     Text for the help hint will be rendered when the same button has a the prop disabled set to true.
 * @param {boolean} [inline]
 * @param {boolean} loading
 *     While true, loading animation will be shown.
 *  @param {boolean} [noSpacing]
 *     Don't include the right margin
 * @param {boolean} active
 *     Active style of the button for when it's being used as a toggle.
 * @param {array} [flags]
 *     Set the flag for "add-button-margin" to override the special add button margin styling.
 * @example
 *  <Button
 *      label="Button"
 *      type="Danger">
 *  </Button>
 *
 */

const buttonTypes = {
    PRIMARY: "primary",
    SECONDARY: "secondary",
    STANDARD: "standard",
    SUCCESS: "success",
    CANCEL: "cancel",
    DANGER: "danger"
};

const buttonTypeList = Object.values(buttonTypes);

const dontFocus = event => event.preventDefault();

const checkFixAddMargin = flags => flags.includes("add-button-margin");

function Button ({
    active,
    alignInputs,
    children,
    className,
    "data-id": dataId,
    disabled,
    disabledText,
    flags,
    href,
    iconName,
    inline,
    label,
    loading,
    noSpacing,
    onClick,
    submit,
    target,
    type
}) {
    const fixAddMargin = checkFixAddMargin(flags);

    if (iconName === "add" && !fixAddMargin && !noSpacing) {
        cannonballChangeWarning({
            message: (
                `Add buttons will no longer automatically have their right margin stripped. ` +
                    `Instead, the 'noSpacing' prop will manually do it. ` +
                    `Switch over by adding the 'add-button-margin' flag to Add buttons.`
            ),
        });
    }

    const classes = classnames(
        "button",
        className,
        iconName,
        {
            "align-inputs": alignInputs,
            "button--add-margin-fix": iconName === "add" && fixAddMargin,
            "button--no-spacing": noSpacing,
            inline,
            loading: loading,
            disabled: disabled,
            "button--active": active,
            "ellipsis-loader-button": loading,
            [type]: buttonTypeList.includes(type)
        }
    );

    const TagName = href ? "a" : "button";

    const Tags = (
        <TagName
            className = {classes}
            data-id={dataId}
            onMouseDown={dontFocus}
            onClick={onClick}
            disabled={disabled}
            type={submit ? "submit" : "button"}
            href={href}
            target={target}
        >
            {label}
            {children}
            <EllipsisLoader loading={loading}/>
        </TagName>
    );

    return (
        disabled && disabledText
            ? <HelpHint
                data-id="button_help-hint"
                hintText={disabledText}
                placement="top"
            >
                {Tags}
            </HelpHint>
            : Tags
    );
}

Button.propTypes = {
    active: PropTypes.bool,
    alignInputs: PropTypes.bool,
    className: PropTypes.string,
    "data-id": PropTypes.string,
    disabled: PropTypes.bool,
    disabledText: PropTypes.string,
    flags: flagsPropType,
    href: PropTypes.string,
    iconName: PropTypes.string,
    inline: PropTypes.bool,
    label: PropTypes.node,
    loading: PropTypes.bool,
    noSpacing: PropTypes.bool,
    onClick: PropTypes.func,
    submit: PropTypes.bool,
    target: PropTypes.string,
    type: PropTypes.oneOf(buttonTypeList),
};

Button.defaultProps = {
    active: false,
    alignInputs: false,
    "data-id": "button",
    disabled: false,
    flags: [],
    loading: false,
    noSpacing: false,
    onClick: () => {},
    submit: false,
};

Button.buttonTypes = buttonTypes;

export default Button;
