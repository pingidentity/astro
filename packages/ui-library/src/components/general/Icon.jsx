"use strict";

import React from "react";
import PropTypes from "prop-types";
import classnames from "classnames";
import { getIconClassName } from "../../util/PropUtils";
import { cannonballChangeWarning } from "../../util/DeprecationUtils";

/**
 * @class Icon
 * @desc A component for displaying an icon by itself or next to next.
 *
 * @param {string} [data-id="icon"]
 *     Defines the "data-id" for top-level HTML container.
 * @param {string} [iconName]
 *     The name of the icon
 * @param {string} [className]
 *     Extra CSS class(s) applied to the top-level HTML container.
 * @param {string} [type]
 *     Set to "leading" or "inline" to either provide extra spacing or not
 *
 * @example
 * <Icon iconName="cog">
 *    Some text or other content that appears next to the icon.
 * </Icon>
 *
 */

const Icon = ({ className, children, "data-id": dataId, type, ...props }) => {
    if (type === "inline") {
        return <span data-id={dataId} className={classnames(getIconClassName(props), className)} />;
    } else if (type !== "leading" && !children) {
        cannonballChangeWarning({
            message: (
                `By default, Icon will display a simple icon when no children are provided. ` +
                `You can use this rendering behavior now by setting the 'type' prop to 'inline'. ` +
                `To display a leading icon that includes the right margin, set 'type' to 'leading'.`
            ),
        });
    }

    return (
        <div className={classnames("icon", className)} data-id={dataId}>
            <div
                className={classnames("icon__graphic", getIconClassName(props))}
                data-id={`${dataId}-graphic`}
            />
            {children && (
                <div className="icon__content" data-id={`${dataId}-content`}>
                    {children}
                </div>
            )}
        </div>
    );
};

Icon.propTypes = {
    "data-id": PropTypes.string,
    className: PropTypes.string,
    iconName: PropTypes.string,
    type: PropTypes.oneOf([ "leading", "inline" ]),
};

Icon.defaultProps = {
    "data-id": "icon",
};

module.exports = Icon;
