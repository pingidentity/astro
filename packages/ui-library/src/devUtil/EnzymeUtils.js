import { mount } from "enzyme";


export const snapshotDataIds = node => {
    const elements = node.find("[data-id]").hostNodes(); // Find only actual HTML elements

    const ids = elements.map(element => element.prop("data-id")).filter(item => item !== "");
    expect(ids).toMatchSnapshot();
};

export const mountSnapshotDataIds = node => snapshotDataIds(mount(node));
