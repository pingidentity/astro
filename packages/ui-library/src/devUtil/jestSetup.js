
// This is needed for testing the modal and DetailsTooltip component.
global.getSelection = jest.fn();
global.getSelection.mockReturnValue({
    toString: () => "",
});
global.document.createRange = () => ({
    setStart: () => {},
    setEnd: () => {},
    commonAncestorContainer: {
        nodeName: "BODY",
        ownerDocument: document,
    },
});
