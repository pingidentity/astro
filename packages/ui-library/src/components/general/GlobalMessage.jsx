import React from "react";
import PropTypes from "prop-types";
import classnames from "classnames";
import { noop } from "underscore";
import Button, { buttonTypes } from "../buttons/Button";
import FlexRow, { alignments, justifyOptions } from "../layout/FlexRow";
import Icon from "./Icon";

/**
 * @enum {string}
 * @alias GlobalMessage.messageTypes
 */
export const messageTypes = {
    /** error */
    ERROR: "error",
    /** info */
    INFO: "info",
    /** warning */
    WARNING: "warning",
};

/**
 * @class GlobalMessage
 * @desc A message meant to appear at the top of the screen- pushes down all application content.
 *       Intended to be used for application expiration messages.
 *
 * @param {string} [data-id]
 *     To define the base "data-id" value for the top-level HTML container.
 * @param {string} [className]
 *     A class name to be applied to the top level of the component.
 * @param {node} [children]
 *     A node that will appear in the component, to the left of the button.
 * @param {node} [buttonLabel]
 *     The label for the button.
 * @param {GlobalMessage.messageTypes} [type="info"]
 *     The type of warning message to show. This controls the background color,
 *     button styling and whether the clear button appears.
 * @param {function} [onButtonClick]
 *     Callback to be triggered when the message button is clicked.
 * @param {function} [onClear]
 *     Callback to be triggered when the X at the right side of the message is clicked.
 */

function GlobalMessage({
    buttonLabel,
    className,
    children,
    "data-id": dataId,
    onButtonClick,
    onClear,
    type
}) {
    return (
        <FlexRow
            className={classnames(
                className,
                "global-message",
                `global-message--${type}`
            )}
            data-id={dataId}
            justify={justifyOptions.CENTER}
        >
            <FlexRow
                alignment={alignments.CENTER}
                className="global-message__content"
            >
                {children}
                {buttonLabel &&
                <Button
                    className={classnames(
                        "global-message__button",
                        `global-message__button--${type}`
                    )}
                    data-id="global-message__button"
                    inline
                    label={buttonLabel}
                    onClick={onButtonClick}
                    type={type === messageTypes.ERROR ? buttonTypes.DANGER : undefined}
                />
                }
            </FlexRow>
            {type !== messageTypes.ERROR &&
                <FlexRow
                    alignment={alignments.CENTER}
                    className="global-message__clear"
                >
                    <Icon
                        data-id="global-message__clear"
                        iconName="clear"
                        onClick={onClear}
                        type="inline"
                    />
                </FlexRow>
            }
        </FlexRow>
    );
}

GlobalMessage.propTypes = {
    buttonLabel: PropTypes.string,
    className: PropTypes.string,
    "data-id": PropTypes.string,
    onButtonClick: PropTypes.func,
    type: PropTypes.oneOf(Object.values(messageTypes))
};

GlobalMessage.defaultProps = {
    "data-id": "global-message",
    onButtonClick: noop,
    type: messageTypes.INFO
};

export default GlobalMessage;
