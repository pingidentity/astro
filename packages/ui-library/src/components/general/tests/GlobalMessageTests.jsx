import React from "react";
import { shallow } from "enzyme";
import { mountSnapshotDataIds } from "../../../devUtil/EnzymeUtils";
import GlobalMessage, { messageTypes } from "../GlobalMessage";
import Button, { buttonTypes } from "../../buttons/Button";
import Icon from "../Icon";

describe("Global Message", () => {
    it("data-id's don't change", () => {
        mountSnapshotDataIds(
            <GlobalMessage
                type={messageTypes.WARNING}
            />
        );
    });

    it("renders the component", () => {
        const component = shallow(
            <GlobalMessage />
        );

        expect(component.exists()).toEqual(true);
    });

    it("renders a clear icon if the type is not set to ERROR", () => {
        const component = shallow(
            <GlobalMessage
                type={messageTypes.WARNING}
            />
        );

        expect(component.find(Icon).exists()).toEqual(true);
    });

    it("does not render a clear icon if the type is set to ERROR", () => {
        const component = shallow(
            <GlobalMessage
                type={messageTypes.ERROR}
            />
        );

        expect(component.find(Icon).exists()).toEqual(false);
    });

    it("renders a button if buttonLabel is passed in", () => {
        const component = shallow(
            <GlobalMessage
                buttonLabel="Really just an incredible button"
            />
        );

        const button = component.find(Button);

        expect(button.exists()).toEqual(true);
        expect(button.props().type).toBeUndefined();
    });

    it("does not render a button if buttonLabel is not passed in", () => {
        const component = shallow(
            <GlobalMessage />
        );

        expect(component.find(Button).exists()).toEqual(false);
    });

    it("renders button with danger type if message type is error", () => {
        const component = shallow(
            <GlobalMessage
                buttonLabel="VERY DANGEROUS PLEASE RUN"
                type={messageTypes.ERROR}
            />
        );

        const scaryButton = component.find(Button);

        expect(scaryButton.props().type).toEqual(buttonTypes.DANGER);
    });
});
