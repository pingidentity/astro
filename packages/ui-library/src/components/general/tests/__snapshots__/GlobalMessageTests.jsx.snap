// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global Message data-id's don't change 1`] = `
Array [
  "global-message",
  "flex-row",
  "flex-row",
  "global-message__clear",
]
`;
